File: Data-Agent-Client/src/services/sqlExecution.service.ts
import http from '../lib/http';
import type { ExecuteSqlResponse, ExecuteSqlParams } from '../types/sql';
import { SqlPaths } from '../constants/apiPaths';

interface ExecuteSqlApiPayload {
  connectionId: number;
  catalog?: string;
  schema?: string;
  sql: string;
}

function toOptionalScope(value?: string | null): string | undefined {
  return value != null && value !== '' ? value : undefined;
}

export const sqlExecutionService = {
  /**
   * Execute SQL on a specific connection and database
   * @param params - ExecuteSqlParams containing connectionId, database, schema, and SQL
   * @returns ExecuteSqlResponse with results or error
   */
  executeSql: async (params: ExecuteSqlParams): Promise<ExecuteSqlResponse> => {
    const payload: ExecuteSqlApiPayload = {
      connectionId: params.connectionId,
      catalog: toOptionalScope(params.databaseName),
      schema: toOptionalScope(params.schemaName),
      sql: params.sql,
    };
    const response = await http.post<ExecuteSqlResponse>(SqlPaths.EXECUTE, payload);
    return response.data;
  },
};

File: Data-Agent-Server/data-agent-server-app/src/main/java/edu/zsc/ai/domain/service/db/SqlExecutionService.java
package edu.zsc.ai.domain.service.db;

import edu.zsc.ai.domain.model.dto.request.db.AgentExecuteSqlRequest;
import edu.zsc.ai.domain.model.dto.response.db.ExecuteSqlResponse;

import java.util.List;

/**
 * Service for executing SQL on a user-owned connection.
 */
public interface SqlExecutionService {

    /**
     * Execute SQL in the context of the given request (connection, database, schema, user).
     *
     * @param request execution context (with userId) and SQL
     * @return execution result (query result set or DML affected rows, or error info)
     */
    ExecuteSqlResponse executeSql(AgentExecuteSqlRequest request);

    /**
     * Execute multiple SQL statements in batch, opening the connection once.
     * Each statement is executed independently — a failure in one does not affect the others.
     *
     * @param connectionId target connection
     * @param databaseName target database (catalog)
     * @param schemaName   target schema (nullable)
     * @param userId       owner user
     * @param sqls         list of SQL statements to execute
     * @return one response per statement, in the same order as input
     */
    List<ExecuteSqlResponse> executeBatchSql(Long connectionId, String databaseName,
                                             String schemaName, Long userId, List<String> sqls);
}
